hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/dn/checksum/TestContainerCommandReconciliation.java

```diff
@@ -252,17 +272,155 @@ public void testGetChecksumInfoSuccess() throws Exception {
     }
   }
 
-  private long writeDataAndGetContainer(boolean close) throws Exception {
+  @Test
+  public void testContainerChecksumWithBlockMissing() throws Exception {
+    // 1. Write data to a container.
+    long containerID = writeDataAndGetContainer(true, 20 * 1024 * 1024);
+    Set<DatanodeDetails> peerNodes = cluster.getHddsDatanodes().stream().map(
+        HddsDatanodeService::getDatanodeDetails).collect(Collectors.toSet());
+    HddsDatanodeService hddsDatanodeService = cluster.getHddsDatanodes().get(0);
+    DatanodeStateMachine datanodeStateMachine = hddsDatanodeService.getDatanodeStateMachine();
+    Container<?> container = datanodeStateMachine.getContainer().getContainerSet().getContainer(containerID);
+    KeyValueContainerData containerData = (KeyValueContainerData) container.getContainerData();
+    ContainerProtos.ContainerChecksumInfo oldContainerChecksumInfo = readChecksumFile(container.getContainerData());
+    KeyValueHandler kvHandler = (KeyValueHandler) datanodeStateMachine.getContainer().getDispatcher()
+        .getHandler(ContainerProtos.ContainerType.KeyValueContainer);
+
+    BlockManager blockManager = kvHandler.getBlockManager();
+    List<BlockData> blockDatas = blockManager.listBlock(container, -1, 100);
+    List<BlockData> deletedBlocks = new ArrayList<>();
+    String chunksPath = container.getContainerData().getChunksPath();
+    long oldDataChecksum = oldContainerChecksumInfo.getContainerMerkleTree().getDataChecksum();
+
+    // 2. Delete some blocks to simulate missing blocks.
+    try (DBHandle db = BlockUtils.getDB(containerData, conf);
+         BatchOperation op = db.getStore().getBatchHandler().initBatchOperation()) {
+      for (int i = 0; i < blockDatas.size(); i += 2) {
+        BlockData blockData = blockDatas.get(i);
+        // Delete the block metadata from the container db
+        db.getStore().getBlockDataTable().deleteWithBatch(op, containerData.getBlockKey(blockData.getLocalID()));
+        // Delete the block file.
+        Files.deleteIfExists(Paths.get(chunksPath + "/" + blockData.getBlockID().getLocalID() + ".block"));
+        deletedBlocks.add(blockData);
+      }
+      db.getStore().getBatchHandler().commitBatchOperation(op);
+      db.getStore().flushDB();
+    }
+
+    Files.deleteIfExists(getContainerChecksumFile(container.getContainerData()).toPath());
+    kvHandler.createContainerMerkleTree(container);
```

Comment:
Add a TODO here to use on demand scanning instead once the scanner integration PR is merged